### ISSUE TITLE
Earth 4

### ISSUE THREAD

**Issue by shacklefordjames60-coder** [mfouad/awesome-egypt-opensource#62]
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from scipy.spatial.distance import hamming
import matplotlib.pyplot as plt
import warnings; warnings.filterwarnings('ignore')

# Updated 2025 Labels with Tool Data
labels_2025 = {
    'Alpha_Cen_A_b': torch.tensor([0.4]),  # 0110
    'GJ_251c': torch.tensor([1.0]),        # 1111 (water hints)
    'TRAPPIST_1e': torch.tensor([0.9]),    # 1111 (atmos hints)
    'HD_219134_b': torch.tensor([0.8])     # 1101
}
target_binaries = {
    'Alpha_Cen_A_b': np.array([0,1,1,0]),
    'GJ_251c': np.array([1,1,1,1]),
    'TRAPPIST_1e': np.array([1,1,1,1]),
    'HD_219134_b': np.array([1,1,0,1])
}

# Multi-Target Dataset Simulation (noisy texts for each)
def encode_noisy_text(text, noise_prob=0.25):
    binary = np.array([1 if c in 'habitablewateratmos' else 0 for c in text.lower()]).astype(float)
    noise = np.random.binomial(1, noise_prob, len(binary))
    return np.pad(binary ^ noise, (0, 128 - len(binary)))

noisy_texts = {
    'Alpha_Cen_A_b': 'Alpha Cen gas giant HZ thick atmos',
    'GJ_251c': 'GJ251c superEarth water HZ 18ly',
    'TRAPPIST_1e': 'TRAPPIST1e atmos hints nitrogen watery',
    'HD_219134_b': 'HD219134b rocky HZ biosig potential'
}
dataset = {k: torch.tensor(encode_noisy_text(v)).float().unsqueeze(0).unsqueeze(0) for k,v in noisy_texts.items()}

# Refined Model (from prior)
class MultiTargetQSNN(nn.Module):
    def __init__(self, seq_len=128):
        super().__init__()
        self.conv = nn.Conv1d(1, 16, 3, padding=1)
        self.attention = nn.MultiheadAttention(16, num_heads=2, batch_first=True)
        self.decoder = nn.Linear(16, 32)
        self.quantum_layer = nn.Linear(32, 32)
        self.dropout = nn.Dropout(0.2)
        self.fc = nn.Linear(32, 1)

    def forward(self, x):
        x = torch.relu(self.conv(x))
        x, _ = self.attention(x, x, x)
        x = x.mean(dim=1)
        x = self.dropout(x)
        decoded = torch.tanh(self.decoder(x))
        x = torch.tanh(self.quantum_layer(decoded))
        out = self.fc(x)
        return out

# Full Multi-Target Training
model = MultiTargetQSNN()
criterion = nn.BCEWithLogitsLoss()
optimizer = optim.Adam(model.parameters(), lr=0.005)
total_loss = 0
for target, tensor in dataset.items():
    label = labels_2025[target]
    binary = target_binaries[target]
    optimizer.zero_grad()
    pred = model(tensor)
    loss = criterion(pred, label)
    loss.backward()
    torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
    optimizer.step()
    total_loss += loss.item()
    with torch.no_grad():
        hab = torch.sigmoid(pred).item()
        conf = 1 - hamming(tensor[0,0,:4].numpy(), binary)
        print(f'{target}: Habitability {hab:.3f}, Confidence {conf:.3f}')

print(f'Avg Loss: {total_loss / len(dataset):.4f}')

# GJ 251c Image Gen
fig, ax = plt.subplots(figsize=(8,6))
ax.set_facecolor('black')
circle = plt.Circle((0,0), 0.1, color='orange', label='GJ 251 (M-dwarf)')
ax.add_patch(circle)
hz_inner = plt.Circle((0,0), 0.15, color='blue', alpha=0.3, fill=False, label='HZ Inner')
hz_outer = plt.Circle((0,0), 0.25, color='blue', alpha=0.3, fill=False, label='HZ Outer')
ax.add_patch(hz_inner); ax.add_patch(hz_outer)
planet = plt.Circle((0.2, 0), 0.03, color='lightblue', label='GJ 251c (Super-Earth)')
ax.add_patch(planet)
ax.text(0.3, 0.1, 'Binary: 1111\n(Water Potential)', fontsize=10, color='white')
ax.set_xlim(-0.5, 0.5); ax.set_ylim(-0.5, 0.5)
ax.set_title('GJ 251c Schematic (18 ly, JWST 2025)', color='white')
ax.legend()
plt.savefig('gj251c_schematic.png')
print('Image saved: gj251c_schematic.png')
